### PR TITLE
fix silent lint ignore for non existing packaged charts

### LINKED ISSUE #6428: Helm v3: lint silently ignores non existing packaged charts
Output of `helm version`:
```
$ helm3 version
version.BuildInfo{Version:"v3.0.0-beta.3", GitCommit:"5cb923eecbe80d1ad76399aee234717c11931d9a", GitTreeState:"clean", GoVersion:"go1.12.9"}

$ helm version --client
Client: &version.Version{SemVer:"v2.14.3", GitCommit:"0e7f3b6637f7af8fcfddb3d2941fcc7cbebb0085", GitTreeState:"clean"}
```
Output of `kubectl version`:
N/A

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
N/A

----

If you try to lint a packaged chart that doesn't exist, helm3 silently ignores it.
 
```
$ helm3 lint aaa-1.0.0.tgz
==> Linting aaa-1.0.0.tgz

1 chart(s) linted, 0 chart(s) failed
```

```
$ helm lint aaa-1.0.0.tgz
==> Skipping aaa-1.0.0.tgz
open aaa-1.0.0.tgz: no such file or directory

0 chart(s) linted, no failures
```

### PR DESCRIPTION
**Pull request by karuppiah7890**
**What this PR does / why we need it**:
Fixes #6428 

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Thanks for the PR! There are a few changes needed here to handle the errors a bit better. I am also wondering if the `lintError` function should set `ErrorSev` as the highest severity [here](https://github.com/helm/helm/pull/6536/files#diff-8254eeb06bb9f2a4e69f4176be9ef319R85) to make sure that any returned error is added to the `result.Errors` slice in `Run`. That would also solve the problem without needed to return `errLintNoChart` for everything because right now, pretty much any error is swallowed

File: pkg/action/lint.go [hunk 1]
```diff
@@ -93,12 +93,12 @@ func lintChart(path string, vals map[string]interface{}, namespace string, stric
...
 		file, err := os.Open(path)
 		if err != nil {
-			return linter, err
+			return linter, errLintNoChart
 		}
 		defer file.Close()
 
 		if err = chartutil.Expand(tempDir, file); err != nil {
-			return linter, err
+			return linter, errLintNoChart
```
[the remark this answers is not in the document]
Comment (karuppiah7890) [reply pkg/action/lint.go]: Yeah, makes sense. I think like you said, let's not swallow these errors. They could be permission issues, corrupted charts and I don't know what other errors. I would say let's not label it too. Let's first bubble it up and fail the lint?

File: pkg/action/lint.go [hunk 2]
```diff
@@ -93,12 +93,12 @@ func lintChart(path string, vals map[string]interface{}, namespace string, stric
 
 		file, err := os.Open(path)
 		if err != nil {
-			return linter, err
+			return linter, errLintNoChart
```
[the remark this answers is not in the document]
Comment (karuppiah7890) [reply pkg/action/lint.go]: You are right. Errors like permission issues could occur. True. So, if other errors could occur, why not just send the whole error itself and then fail the lint? "Not exist" is just one of the errors. Or do we want to handle that separately? However I think the error will show up in the console

**Comment by karuppiah7890** [comment]
@thomastaylor312 So, what do you feel? Do you want to specifically handle errors that a chart is corrupted / chart is missing and then just bubble up other errors as is and fail lint?

**Comment by karuppiah7890** [comment]
If we bubble out errors, it's not so bad for some of them. Some examples of permission denied and EOF errors. I'm not sure in what cases EOF occurs. One I can think of is, a `.tgz` not really being a tgz, which is what I did. I just did `touch meh.tgz`. Another is a corrupted one - like only half the content is downloaded from Internet or some place? It's a guess from this error - https://github.com/helm/helm/issues/6140

```
$ helm3 lint blah-0.1.0.tgz
==> Linting blah-0.1.0.tgz
Error open blah-0.1.0.tgz: permission denied

Error: 1 chart(s) linted, 1 chart(s) failed

$ helm3 lint dummy-chart/
==> Linting dummy-chart/
Error stat dummy-chart/Chart.yaml: permission denied

Error: 1 chart(s) linted, 1 chart(s) failed

$ helm3 lint meh.tgz
==> Linting meh.tgz
Error EOF

Error: 1 chart(s) linted, 1 chart(s) failed
```

**Comment by karuppiah7890** [comment]
If we just bubble up, I think we don't need a `errLintNoChart` variable as it'll not be used. May be instead of blindly bubbling up errors, we can say what we tried to do and what the error was. Like `opening tgz file: open blah-0.1.0.tgz: permission denied`, `extracting tgz file: EOF`

**Comment by karuppiah7890** [comment]
I'll also make these changes to #6537 PR

**Comment by karuppiah7890** [comment]
There are actually so many kinds of errors that occur in functions like [`chartutil.Expand`](https://github.com/helm/helm/blob/4f26b658d867569ba69db9324be64d3dfdee2440/pkg/chartutil/expand.go#L28)

**Comment by karuppiah7890** [comment]
I have removed the `lowestTolerance` as it was only being used when there were errors returned by `lintChart` and every error `lintChart` returned was actually a valid and high severity. There was no possibility for warnings or infos over there, as we are clearly returning an error. If it was some warning or info or some unknown, we would put it in linter messages and returned nil error.

So, got rid of the extra code for when error occurs and wrapped the errors with some meaningful messages to understand the error. Let me know if it makes sense. I'll also see if I need to add more tests!

File: pkg/action/lint.go [hunk 3]
```diff
@@ -87,18 +76,18 @@ func lintChart(path string, vals map[string]interface{}, namespace string, stric
 	if strings.HasSuffix(path, ".tgz") {
 		tempDir, err := ioutil.TempDir("", "helm-lint")
 		if err != nil {
-			return linter, err
+			return linter, errors.Errorf("creating temp dir to extract tarball: %s", err.Error())
```
Comment (thomastaylor312) [inline pkg/action/lint.go]: All of these should just be `errors.Wrap(err, "message")`

Comment (thomastaylor312) [reply pkg/action/lint.go]: Nit: The errors should probably be more clear with a message like "unable to create temp dir..."

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
So I think this is looking good, just a few small changes to the errors. I would like another maintainer who has a bit more experience with the lint code double check that this change won't have a larger impact

File: pkg/action/lint.go [hunk 4]
```diff
@@ -52,28 +50,19 @@ func NewLint() *Lint {
...
-				}
+			result.Errors = append(result.Errors, err)
+			continue
+		}
+
+		result.Messages = append(result.Messages, linter.Messages...)
+		result.TotalChartsLinted++
+		for _, msg := range linter.Messages {
+			if msg.Severity == support.ErrorSev {
+				result.Errors = append(result.Errors, msg.Err)
```
Comment (mattfarina) [inline pkg/action/lint.go]: I think we might have a bug that was in the old code that was carried over to the new logic. When `l.Strict` is false `support.ErrorSev` is the error level. When `l.Strict` is true `support.WarningSev` should be used for appending to `result.Errors`.

This is important for the `helm lint` command as the different levels tell it when to return a non-zero exit code to the OS.

In the old code `lowestTolerance` was doing this when an err was returned from `lintChart`.  Could you fix this logic?

For example...
```
lowestTolerance := support.ErrorSev
if l.Strict {
	lowestTolerance = support.WarningSev
}
```

and then later...
```
if msg.Severity >= lowestTolerance {
```

**Review by mattfarina — CHANGES_REQUESTED** [review]
There appears to be one small bug in the code around strict handling

File: pkg/action/lint.go (hunk 4 above)
Comment (karuppiah7890) [reply pkg/action/lint.go]: Ah sure. I'll add some failing tests and do it. And I did doubt this, but I really didn't know how lint was entirely working. I had read some code inside lint.All and sorts, to understand things like highest severity, not much at the higher level. I better read more code 😅 

So this way we will also be getting warnings along with errors in strict case but only errors in a normal (non strict) case

**Comment by karuppiah7890** [comment]
Thanks for the reviews @thomastaylor312 and @mattfarina , I have incorporated them in my PR 😄 please do check it out!

**Review by thomastaylor312 — APPROVED** [review]
This looks good! Thank you for the unit tests here!

**Pull request merged**